File: src/main/java/org/eclairjs/nashorn/wrap/mllib/linalg/Vector.java
package org.eclairjs.nashorn.wrap.mllib.linalg;
/*                                                                         
* Copyright 2016 IBM Corp.                                                 
*                                                                          
* Licensed under the Apache License, Version 2.0 (the "License");          
* you may not use this file except in compliance with the License.         
* You may obtain a copy of the License at                                  
*                                                                          
*      http://www.apache.org/licenses/LICENSE-2.0                          
*                                                                          
* Unless required by applicable law or agreed to in writing, software      
* distributed under the License is distributed on an "AS IS" BASIS,        
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
* See the License for the specific language governing permissions and      
* limitations under the License.                                           
*/ 

import org.eclairjs.nashorn.Utils;
import org.eclairjs.nashorn.wrap.WrappedFunction;
import org.eclairjs.nashorn.wrap.WrappedClass;


public class Vector extends WrappedClass {

    static WrappedFunction F_size = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            Object returnValue;
            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
            returnValue = _vector.size();
            return returnValue;
        }
    };

    static WrappedFunction F_toArray = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            Object returnValue;
            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
            returnValue = Utils.javaToJs(_vector.toArray()); // var ret = Java.from(a); // convert java array to JavaScript array.
            return returnValue;
        }
    };

    static WrappedFunction F_equals = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            Object returnValue;
            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
            Object other = Utils.jsToJava(args[0]);
            returnValue = _vector.equals(other);
            return returnValue;
        }
    };

    static WrappedFunction F_hashCode = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            Object returnValue;
            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
            returnValue = _vector.hashCode();
            return returnValue;
        }
    };

    static WrappedFunction F_apply = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            Object returnValue;
            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
            int i = (int) args[0];
            returnValue = _vector.apply(i);
            return returnValue;
        }
    };

    static WrappedFunction F_copy = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            Object returnValue ;
            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
            returnValue = _vector.copy();
            return Utils.javaToJs(returnValue);
        }
    };

//    static WrappedFunction F_foreachActive = new WrappedFunction() {
//        @Override
//        public Object call(Object thiz, Object... args) {
//            Object returnValue = null;
//            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
//            Object  bindArgs0 = null;
//            if (args.length > 1) {
//                bindArgs = args[1];
//            }
//            JSFunction2 f = (JSFunction2)Utils.createLambdaFunction(args[0], "org.eclairjs.nashorn.JSFunction2", _vector.context(), bindArgs0);
//            _vector.foreachActive(f);
//            return null;
//        }
//    };

    static WrappedFunction F_numActives = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            Object returnValue;
            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
            returnValue = _vector.numActives();
            return returnValue;
        }
    };

    static WrappedFunction F_numNonzeros = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            Object returnValue;
            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
            returnValue = _vector.numNonzeros();
            return returnValue;
        }
    };

    static WrappedFunction F_toSparse = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            Object returnValue;
            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
            returnValue = _vector.toSparse();
            return Utils.javaToJs(returnValue);
        }
    };

    static WrappedFunction F_toDense = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            Object returnValue;
            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
            returnValue = _vector.toDense();
            return Utils.javaToJs(returnValue);
        }
    };

    static WrappedFunction F_compressed = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            Object returnValue;
            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
            returnValue = _vector.compressed();
            return Utils.javaToJs(returnValue);
        }
    };

    static WrappedFunction F_argmax = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            Object returnValue;
            org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector) ((Vector) thiz).getJavaObject();
            returnValue = _vector.argmax();
            return returnValue;
        }
    };


    private org.apache.spark.mllib.linalg.Vector _vector;

    public Vector(org.apache.spark.mllib.linalg.Vector _vector)
    { this._vector = _vector; }

    public Vector() {

    }

    static public String getModuleName() {
        return "mllib.linalg.Vector";
    }

    public boolean checkInstance(Object other) {
        return other instanceof Vector;
    }

    public Object getJavaObject() {
        return _vector;
    }

    @Override
    public String toString() {

        return this.getJavaObject().toString();
    }

    @Override
    public String toJSON() {

        String returnValue;
        org.apache.spark.mllib.linalg.Vector _vector = (org.apache.spark.mllib.linalg.Vector)  this.getJavaObject();
        returnValue = _vector.toJson();
        return returnValue;
    }
    public String getClassName() {
        return "Vector";
    }

    // get the value of that named property
    @Override
    public Object getMember(String name) {
        switch (name) {
            case "size":
                return F_size;
            case "toArray":
                return F_toArray;
            case "equals":
                return F_equals;
            case "hashCode":
                return F_hashCode;
            case "apply":
                return F_apply;
            case "copy":
                return F_copy;
//            case "foreachActive":
//                return F_foreachActive;
            case "numActives":
                return F_numActives;
            case "numNonzeros":
                return F_numNonzeros;
            case "toSparse":
                return F_toSparse;
            case "toDense":
                return F_toDense;
            case "compressed":
                return F_compressed;
            case "argmax":
                return F_argmax;

        }
        return super.getMember(name);
    }

    @Override
    public boolean hasMember(String name) {
        switch (name) {
            case "$init$":
            case "size":
            case "toArray":
            case "equals":
            case "hashCode":
            case "apply":
            case "copy":
//            case "foreachActive":
            case "numActives":
            case "numNonzeros":
            case "toSparse":
            case "toDense":
            case "compressed":
            case "argmax":
                return true;
        }
        return super.hasMember(name);
    }

}

File: server/src/main/resources/eclairjs/mllib/linalg/Vector.js
/*
 * Copyright 2016 IBM Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

(function () {
    //var JavaWrapper = require(EclairJS_Globals.NAMESPACE + '/JavaWrapper');
    //var Logger = require(EclairJS_Globals.NAMESPACE + '/Logger');
    //var Utils = require(EclairJS_Globals.NAMESPACE + '/Utils');
    /**
     * Represents a numeric vector, whose index type is Int and value type is Double.
     *
     * Note: Users should not implement this interface.
     * @classdesc
     * @constructor Vector
     * @abstract
     * @memberof module:eclairjs/mllib/linalg
     */

    var Vector = Java.type('org.eclairjs.nashorn.wrap.mllib.linalg.DenseVector');
    //var Vector = function (jvmObject) {
    //
    //    this.logger = Logger.getLogger("Vector_js");
    //    JavaWrapper.call(this, jvmObject);
    //
    //};
    //
    //Vector.prototype = Object.create(JavaWrapper.prototype);
    //
    //Vector.prototype.constructor = Vector;



    /**
     * Size of the vector.
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#size
     * @returns {integer}
     */
    //Vector.prototype.size = function () {
    //    return this.getJavaObject().size();
    //};


    /**
     * Converts the instance to a double array.
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#toArray
     * @returns {float[]}
     */
    //Vector.prototype.toArray = function () {
    //    var a = this.getJavaObject().toArray();
    //    var ret = Java.from(a); // convert java array to JavaScript array.
    //    return ret;
    //};

    /**
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#equals
     * @param {object} other
     * @returns {boolean}
     */
    //Vector.prototype.equals = function (other) {
    //    var other_uw = Utils.unwrapObject(other);
    //    return this.getJavaObject().equals(other_uw);
    //};


    /**
     *
     * Returns a hash code value for the vector. The hash code is based on its size and its first 128
     * nonzero entries, using a hash algorithm similar to {@link hashCode}.
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#hashCode
     * @returns {number}
     */
    //Vector.prototype.hashCode = function () {
    //    return  this.getJavaObject().hashCode();
    //};


    /**
     * Gets the value of the ith element.
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#apply
     * @param {number} i  index
     * @returns {number}
     */
    //Vector.prototype.apply = function (i) {
    //    return  this.getJavaObject().apply(i);
    //};


    /**
     * Makes a deep copy of this vector.
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#copy
     * @returns {module:eclairjs/mllib/linalg.Vector}
     */
    //Vector.prototype.copy = function () {
    //    var javaObject =  this.getJavaObject().copy();
    //    return Utils.javaToJs(javaObject);
    //};

    /**
     * Number of active entries.  An "active entry" is an element which is explicitly stored,
     * regardless of its value.  Note that inactive entries have value 0.
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#numActives
     * @returns {integer}
     */
    //Vector.prototype.numActives = function () {
    //       return  this.getJavaObject().numActives();
    //};


    /**
     * Number of nonzero elements. This scans all active values and count nonzeros.
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#numNonzeros
     * @returns {integer}
     */
    //Vector.prototype.numNonzeros = function () {
    //       return  this.getJavaObject().numNonzeros();
    //};


    /**
     * Converts this vector to a sparse vector with all explicit zeros removed.
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#toSparse
     * @returns {module:eclairjs/mllib/linalg.SparseVector}
     */
    //Vector.prototype.toSparse = function () {
    //    var javaObject =  this.getJavaObject().toSparse();
    //    return Utils.javaToJs(javaObject);
    //};


    /**
     * Converts this vector to a dense vector.
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#toDense
     * @returns {module:eclairjs/mllib/linalg.DenseVector}
     */
    //Vector.prototype.toDense = function () {
    //    var javaObject =  this.getJavaObject().toDense();
    //    return Utils.javaToJs(javaObject);
    //};


    /**
     * Returns a vector in either dense or sparse format, whichever uses less storage.
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#compressed
     * @returns {module:eclairjs/mllib/linalg.Vector}
     */
    //Vector.prototype.compressed = function () {
    //    var javaObject =  this.getJavaObject().compressed();
    //    return Utils.javaToJs(javaObject);
    //};


    /**
     * Find the index of a maximal element.  Returns the first maximal element in case of a tie.
     * Returns -1 if vector has length 0.
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#argmax
     * @returns {integer}
     */
    //Vector.prototype.argmax = function () {
    //    return  this.getJavaObject().argmax();
    //};


    /**
     * Converts the vector to a JSON string.
     * @function
     * @name module:eclairjs/mllib/linalg.Vector#toJSON
     * @returns {Object}
     * @ignore
     */
    //Vector.prototype.toJSON = function () {
    //    return  this.toArray();
    //};


    module.exports = Vector;

})();
